File: backend/app/schemas/kg.py
"""知识图谱 API Schema — `/kg/*` 请求/响应模型。"""
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class GraphNode(BaseModel):
    """图节点。"""
    id: str = Field(description="节点唯一标识")
    type: Literal["Dataset", "Publisher", "Theme", "Keyword", "Format"] = Field(description="节点类型")
    label: str = Field(description="显示标签")
    source_field: Optional[str] = Field(default=None, description="来源元数据字段")


class GraphEdge(BaseModel):
    """图边。"""
    source: str = Field(description="源节点 id")
    target: str = Field(description="目标节点 id")
    type: str = Field(description="关系类型")
    weight: float = Field(default=1.0, description="权重")


class GraphResponse(BaseModel):
    """图数据响应。"""
    nodes: list[GraphNode]
    edges: list[GraphEdge]


class KgBuildResult(BaseModel):
    """图谱构建结果。"""
    datasets_processed: int = Field(description="处理的 datasets 数")
    entities_upserted: int = Field(description="upsert 的实体数")
    relationships_created: int = Field(description="创建的关系数")
    similar_edges: int = Field(description="Jaccard 相似边数")
    errors: list[str] = Field(default_factory=list, description="处理中的错误列表")


class EntityListItem(BaseModel):
    """实体列表项。"""
    id: str = Field(description="实体 ID（Label:name 格式）")
    type: str = Field(description="实体类型")
    name: str = Field(description="实体名称")
    dataset_count: int = Field(description="关联数据集数量")


class EntityListResponse(BaseModel):
    """实体列表响应。"""
    items: list[EntityListItem]
    page: int
    size: int
    count: int


class EntityDetail(BaseModel):
    """实体详情。"""
    id: str
    type: str
    name: str
    related_datasets: list[dict[str, Any]] = Field(
        default_factory=list,
        description="关联数据集列表",
    )

File: web/src/api/kg.ts
// 知识图谱 API 封装（Phase 5）
// 与后端 /kg/* 路由一一对应

/** 图节点 */
export interface GraphNode {
  id: string
  type: 'Dataset' | 'Publisher' | 'Theme' | 'Keyword' | 'Format'
  label: string
  source_field?: string | null
}

/** 图边 */
export interface GraphEdge {
  source: string
  target: string
  type: string
  weight: number
}

/** 图数据响应 */
export interface GraphResponse {
  nodes: GraphNode[]
  edges: GraphEdge[]
}

/** 图谱构建结果 */
export interface KgBuildResult {
  datasets_processed: number
  entities_upserted: number
  relationships_created: number
  similar_edges: number
  errors: string[]
}

/** 实体列表项 */
export interface EntityListItem {
  id: string
  type: string
  name: string
  dataset_count: number
}

/** 实体列表响应 */
export interface EntityListResponse {
  items: EntityListItem[]
  page: number
  size: number
  count: number
}

/** 实体详情 */
export interface EntityDetail {
  id: string
  type: string
  name: string
  related_datasets: Array<{
    dataset_id: string
    rel_type: string
    confidence: number | null
  }>
}

/** ResponseModel 通用包装 */
interface ApiEnvelope<T> {
  code: number
  data: T | null
  msg: string
}

const BASE = '/api/kg'

async function unwrap<T>(res: Response): Promise<T> {
  const json = (await res.json()) as ApiEnvelope<T>
  if (!res.ok || json.code !== 200 || json.data === null) {
    throw new Error(json.msg || `HTTP ${res.status}`)
  }
  return json.data
}

/** 全量重建知识图谱 */
export async function buildGraph(): Promise<KgBuildResult> {
  const res = await fetch(`${BASE}/build`, { method: 'POST' })
  return unwrap<KgBuildResult>(res)
}

/** 获取图数据 */
export async function getGraph(entityTypes?: string[]): Promise<GraphResponse> {
  const qs = new URLSearchParams()
  if (entityTypes && entityTypes.length > 0) {
    qs.set('entity_types', entityTypes.join(','))
  }
  const url = qs.toString() ? `${BASE}/graph?${qs.toString()}` : `${BASE}/graph`
  const res = await fetch(url)
  return unwrap<GraphResponse>(res)
}

/** 分页实体列表 */
export async function listEntities(
  type?: string,
  page = 1,
  size = 20,
): Promise<EntityListResponse> {
  const qs = new URLSearchParams({ page: String(page), size: String(size) })
  if (type) qs.set('type', type)
  const res = await fetch(`${BASE}/entities?${qs.toString()}`)
  return unwrap<EntityListResponse>(res)
}

/** 实体详情 */
export async function getEntityDetail(entityId: string): Promise<EntityDetail> {
  const res = await fetch(`${BASE}/entities/${encodeURIComponent(entityId)}`)
  return unwrap<EntityDetail>(res)
}

/** 数据集的所有关系 */
export async function getDatasetRelations(datasetId: string): Promise<GraphResponse> {
  const res = await fetch(`${BASE}/datasets/${datasetId}/relations`)
  return unwrap<GraphResponse>(res)
}
